src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -2699,4 +2741,97 @@ void HwEmShim::closemMessengerThread() {
 }
 /********************************************** QDMA APIs IMPLEMENTATION END**********************************************/
 /**********************************************HAL2 API's END HERE **********************************************/
+
+/********************************************** Q2H_helper class implementation starts **********************************************/
+bool device2xrt_trans_cb(bool, int, unsigned long int, void*,unsigned long int) {
+    std::cout << " device2xrt_trans_cb() is called...." <<std::endl;
+    return true;
+}
+Q2H_helper :: Q2H_helper() : buf_size(0),buf(NULL), r(0) {
+    std::cout << " Start of Q2H_helper()..." << std::endl;
+    ci_msg = new call_packet_info;
+    ri_msg = new response_packet_info;
+    Q2h_sock = new unix_socket("xcl_sock_K2H");
+    ci_msg->set_size(0);
+    ci_msg->set_xcl_api(0);
+    ri_msg->set_size(0);
+    ri_msg->set_xcl_api(0);
+    i_len = ci_msg->ByteSize();
+    i_buf = malloc(i_len);
+    ri_len = ri_msg->ByteSize();
+    ri_buf = malloc(i_len);
+    std::cout << " End of Q2H_helper()..." << std::endl;
+}
+Q2H_helper::~Q2H_helper() {
+    delete ci_msg;
+    delete ri_msg;
+}
+void Q2H_helper::poolingon_Qdma() {
+    //Pooling on socket for any memory or interrupt requests from SIM_QDMA
+    r = Q2h_sock->sk_read(i_buf, i_len);
+    if (r <= 0) {
+    	return;
+    }
+    assert(i_len == (uint32_t)r);
+    ci_msg->ParseFromArray(i_buf, i_len);
+    if (buf_size == 0) {
+    	buf = malloc(ci_msg->size());
```

Comment:
same buf pointer is allocated multiple times

Reply:
buf gets set by memset() for every transaction and its get re-allocated with realloc() when size is not sufficient and keeps note of the new size

Reply:
Same comment as earlier.  looks unsafe

Reply:
Smart pointer are used to manage buffer allocation and de-allocation.